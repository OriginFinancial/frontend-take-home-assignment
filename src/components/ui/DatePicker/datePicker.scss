@import '~/assets/styles/_colors.scss';
@import '~/assets/styles/_globalVars.scss';

.datePicker {
  display: flex;
  flex-direction: column;

  & button {
    letter-spacing: -0.01em;
    height: 100%;

    & img {
      width: calc(((#{$space} / 35) + 7px) * 0.5);
      transition: all 0.2s ease-out;
    }

    &:active {
      & img {
        transform: scale(0.8);
      }
    }
  }

  & .back {
    flex: 20;
    background: $color_light_hard;

    &:disabled {
      background: $color_light;
      & img {
        opacity: 0.3;
      }
    }
  }
  & .forward {
    flex: 20;
    background: $color_light_hard;
    transform: rotate(180deg);
  }
  & .pickerDisplay {
    flex: 60;
    flex-flow: column;
    line-height: line-heigth;
    position: relative;

    &:focus {
      box-shadow: inset 0 0 0.5vh rgba($color_primary_soft, 0.15);
    }

    & p {
      font-size: calc(#{$p} * 1.2);
      text-transform: capitalize;
      line-height: 1.33em;

      & small {
        font-weight: normal;
        font-size: 0.7em;
        line-height: 0.7em;
      }
    }
  }

  & .innerHolder {
    display: flex;
    align-items: center;
    justify-content: space-between;
    position: relative;
    width: 100%;
    height: 6vh;
    margin-top: 0.5vh;
    border: $border;
    border-radius: calc(#{$radius} * 2);
  }

  & .picker {
    position: absolute;
    top: -1px;
    left: -1px;
    z-index: 2;
    width: calc(100% + 2px);
    background: $color_light_soft;
    border: $border;
    border-radius: calc(#{$radius} * 2);
    // transition: all 0.2s ease-in-out;
    animation: fade 0.25s ease-in-out;
    animation-iteration-count: 1;

    & .row {
      height: 6vh;

      & .back,
      .forward {
        background: transparent;
      }

      & .pickerDisplay {
        cursor: default;

        & p {
          color: $color_primary_soft;
        }
      }
    }

    & ul {
      display: flex;
      flex-wrap: wrap;
      padding: 1vh 2vw 2.5vh;
      box-shadow: 0 1vh 3vh rgba(black, 0.1);
      z-index: 1;

      & li {
        user-select: none;
        cursor: pointer;
        width: 33.3%;
        text-align: center;
        padding: 0.75vh 0;
        border-radius: 100px;
        transition: all 0.15s ease-in-out;

        &:hover {
          background: $color_primary;

          & p {
            color: $color_light_soft;
          }
        }

        &:active {
          background: $color_primary_soft;
          color: $color_light_soft;

          & p {
            color: $color_light_soft;
          }
        }
      }

      & .disabled {
        cursor: default;
        & p {
          color: $color_dark_soft;
        }

        &:hover {
          background: inherit;

          & p {
            color: $color_dark_soft;
          }
        }
      }

      & .current {
        background: $color_primary_soft;

        & p {
          color: $color_light_soft;
        }
      }
    }
  }
}

@media all and (min-width: 280px) and (max-width: 679px) and (orientation: portrait) {
  .datePicker {
    & button {
      & img {
        width: calc(((#{$space} / 35) + 7px) * 0.75);
      }
    }

    & .back,
    .forward {
      flex: 15;
    }

    & .pickerDisplay {
      flex: 70;

      & p {
        font-size: calc(#{$p} * 1.4);
      }
    }

    & .innerHolder {
      height: 7vh;
    }

    & .picker {
      position: absolute;

      & .row {
        height: 7vh;
      }

      & ul {
        padding: 2vh 10vw 3.5vh;

        & li {
          width: 25%;
          padding: 1.5vh 0;
        }
      }
    }
  }
}

@media all and (min-height: 280px) and (max-height: 679px) and (orientation: landscape) {
  .datePicker {
    & .back,
    .forward {
      flex: 15;
    }

    & .innerHolder {
      height: 7vw;
    }

    & .picker {
      & .row {
        height: 7vw;
      }

      & ul {
        & li {
          padding: 0.75vw 0;
        }
      }
    }
  }
}
