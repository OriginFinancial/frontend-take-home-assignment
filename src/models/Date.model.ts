export interface MonthYear {
  month: string;
  year: number;
}
