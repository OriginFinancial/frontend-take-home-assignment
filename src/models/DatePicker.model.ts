import { KeyboardEvent, MonthYear } from '.';

export interface DatePickerProps {
  data: Data;
  actions: Actions;
}

interface Data {
  selectedDate: MonthYear;
  isPrevButtonDisabled: boolean;
}

interface Actions {
  handlePressNextDate: () => void;
  handlePressPreviousDate: () => void;
  handlePressKeyboardArrows: (e: KeyboardEvent) => void;
}
