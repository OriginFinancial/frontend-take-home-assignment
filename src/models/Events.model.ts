export interface KeyboardEvent {
  keyCode: number;
}

export interface FormEvent {
  e: React.FormEvent<HTMLInputElement>;
}
