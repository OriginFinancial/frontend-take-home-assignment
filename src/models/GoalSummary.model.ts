import { MonthYear } from '.';
export interface GoalSummaryProps {
  data: Data;
}

interface Data {
  amount: string;
  selectedDate: MonthYear;
  totalOfMonths: number;
  monthlyAmount: string;
}
