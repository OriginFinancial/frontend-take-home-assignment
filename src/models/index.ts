export * from './Date.model';
export * from './DatePicker.model';
export * from './Events.model';
export * from './GoalSummary.model';
export * from './SavingGoalForm.model';
export * from './Paper.model';
export * from './TextInput.model';
