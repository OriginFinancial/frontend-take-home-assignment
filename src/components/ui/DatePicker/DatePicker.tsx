import * as React from 'react';
import './datePicker.scss';

const DatePicker: React.FC = () => {
  return (
    <label className="datePicker">
      Reach goal by
      <span className="innerHolder">
        <button className="back">
          <img src={require('~/assets/icons/arrow.svg')} />
        </button>
        <button className="pickerDisplay">
          <h4>October</h4>
          <p>2020</p>
        </button>
        <button className="forward">
          <img src={require('~/assets/icons/arrow.svg')} />
        </button>
      </span>
    </label>
  );
};

export default DatePicker;
