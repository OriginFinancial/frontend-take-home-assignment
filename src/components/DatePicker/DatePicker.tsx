import * as React from 'react';
import styled, { css } from 'styled-components';
import { DatePickerProps } from '../../models';
import { form } from '../../styles';
import ArrowIcon from '../../icons/arrow.svg';

const DatePicker: React.FunctionComponent<DatePickerProps> = ({
  data: { selectedDate, isPrevButtonDisabled },
  actions: {
    handlePressNextDate,
    handlePressPreviousDate,
    handlePressKeyboardArrows
  }
}) => {
  return (
    <Container>
      <IconButton
        onClick={handlePressPreviousDate}
        onKeyDown={handlePressKeyboardArrows}
        disabled={isPrevButtonDisabled}
      >
        <ArrowLeftIcon src={ArrowIcon} />
      </IconButton>
      <DateInfo>
        <Month id="testDateMonth">{selectedDate?.month}</Month>
        <Year id="testDateYear">{selectedDate?.year}</Year>
      </DateInfo>
      <IconButton
        onClick={handlePressNextDate}
        onKeyDown={handlePressKeyboardArrows}
      >
        <ArrowRightIcon src={ArrowIcon} />
      </IconButton>
    </Container>
  );
};

const Container = styled.div`
  align-items: center;
  display: flex;
`;

const Month = styled.b``;

const Year = styled.span``;

const DateInfo = styled.p`
  align-items: center;
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  justify-content: center;
  margin: 0;
  padding: 0.8rem 1.6rem;
`;

const IconButton = styled.button<{ disabled?: boolean }>`
  align-items: center;
  background-color: ${props => props?.theme?.colors?.primary?.extralight};
  border: 0.1rem solid ${props => props?.theme?.colors?.secondary?.main};
  border-radius: 0.3rem;
  cursor: pointer;
  display: flex;
  padding: 1.6rem;
  outline: none;

  ${props =>
    props.disabled &&
    css`
      background: ${props => props?.theme?.colors?.white};
    `}
  ${form.formComponentFocus};
`;
const ArrowLeftIcon = styled.img`
  height: 1.2rem;
  width: 1.2rem;
`;

const ArrowRightIcon = styled.img`
  height: 1.2rem;
  width: 1.2rem;
  transform: rotate(180deg);
`;

export default DatePicker;
