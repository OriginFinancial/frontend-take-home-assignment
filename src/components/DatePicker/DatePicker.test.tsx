import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { DatePickerProps } from '../../models';
import DatePicker from './DatePicker';

describe('DatePicker component', () => {
  let component: ReactWrapper;
  let props: DatePickerProps;

  beforeEach(() => {
    jest.clearAllMocks();
    const fn = jest.fn();
    props = {
      data: {
        selectedDate: {
          month: 'August',
          year: 2020
        },
        isPrevButtonDisabled: false
      },
      actions: {
        handlePressNextDate: fn,
        handlePressPreviousDate: fn,
        handlePressKeyboardArrows: fn
      }
    };
    component = mount(<DatePicker {...props} />);
  });

  it('renders initial layout', () => {
    expect(component.getElements()).toMatchSnapshot();
  });

  it('renders the month prop', () => {
    expect(
      component
        .find('#testDateMonth')
        .first()
        .text()
    ).toBe(props.data.selectedDate.month);
  });

  it('renders the year prop', () => {
    expect(
      component
        .find('span#testDateYear')
        .first()
        .text()
    ).toBe(`${props.data.selectedDate.year}`);
  });
});
