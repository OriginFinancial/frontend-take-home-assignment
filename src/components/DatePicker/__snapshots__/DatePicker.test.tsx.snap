// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker component renders initial layout 1`] = `
Array [
  <DatePicker
    actions={
      Object {
        "handlePressKeyboardArrows": [MockFunction],
        "handlePressNextDate": [MockFunction],
        "handlePressPreviousDate": [MockFunction],
      }
    }
    data={
      Object {
        "isPrevButtonDisabled": false,
        "selectedDate": Object {
          "month": "August",
          "year": 2020,
        },
      }
    }
  />,
]
`;
